crates/astria-sequencer-relayer/src/relayer/read.rs

```diff
@@ -0,0 +1,218 @@
+//! A stream of sequencer blocks.
+use std::{
+    future::Future as _,
+    pin::Pin,
+    sync::Arc,
+    task::Poll,
+    time::Duration,
+};
+
+use astria_eyre::eyre::{
+    self,
+    Report,
+    WrapErr as _,
+};
+use futures::{
+    future::BoxFuture,
+    ready,
+    FutureExt as _,
+};
+use pin_project_lite::pin_project;
+use sequencer_client::{
+    tendermint::block::Height,
+    HttpClient,
+    SequencerBlock,
+};
+use tokio_stream::Stream;
+use tracing::{
+    debug,
+    instrument,
+    warn,
+    Instrument as _,
+    Span,
+};
+
+#[derive(Default, Debug)]
+struct Heights {
+    last_observed: Option<Height>,
+    last_requested: Option<Height>,
+}
+
+impl Heights {
+    /// Returns the next height to be fetched.
+    ///
+    /// If `last_observed` is unset, returns `None`.
+    /// If `last_requested` is unset, returns `last_observed`.
+    /// If both are set, returns `last_requested + 1` if less than
+    /// `last_observed`, `None` otherwise.
+    fn next_height_to_fetch(&self) -> Option<Height> {
+        let last_observed = self.last_observed?;
+        match self.last_requested {
+            None => Some(last_observed),
+            Some(last_requested) if last_requested < last_observed => {
+                Some(last_requested.increment())
+            }
+            Some(..) => None,
+        }
+    }
+}
+
+pin_project! {
+    pub(super) struct BlockStream {
+        client: HttpClient,
+        heights: Heights,
+        #[pin]
+        future: Option<BoxFuture<'static, eyre::Result<SequencerBlock>>>,
+        paused: bool,
+        block_time: Duration,
+        state: Arc<super::State>,
+    }
+}
+
+impl BlockStream {
+    pub(super) fn new(client: HttpClient, state: Arc<super::State>) -> Self {
+        Self {
+            client,
+            heights: Heights::default(),
+            future: None,
+            block_time: Duration::from_millis(1_000),
+            paused: false,
+            state,
+        }
+    }
+
+    pub(super) fn set_block_time(&mut self, block_time: Duration) {
+        self.block_time = block_time;
+    }
+
+    pub(super) fn set_latest_sequencer_height(&mut self, height: Height) {
+        self.heights.last_observed.replace(height);
+    }
+
+    pub(super) fn pause(&mut self) {
+        debug!("stream paused");
```

Comment:
```suggestion
        debug!("block stream paused");
```

same for resume

Reply:
Added and also moved the debug logs out of the methods to the places that actually call the methods.